### DISCUSSION TITLE
Ideas and deprecations for next major release 4.0

### DISCUSSION THREAD

**Discussion by ramunasd** [php-amqplib/php-amqplib#894] [Ideas]
Feel free to comment and suggest what can be changed/refactored to make things more fluent and consistent in next major version.

- [ ] Use one immutable object for all connection options and reuse it in all connection types
- [ ] Merge StreamIO and SocketIO into corresponding connection class
- [ ] Connection class should not extend channel class therefore no calls to protected methods
- [ ] Split AMQPChannel->wait() into to separate blocking and non-blocking methods, separate method for internal use
- [ ] Refactor existing TODO
- [ ] Deprecate AMQPSSLConnection, activate SSL context whenever protocol is secure
- [ ] PSR-12 code style
- [ ] Changelog standard https://keepachangelog.com/en/1.0.0/
- [ ] PHP > 7.0, typehints for most public methods
- [ ] Use https://github.com/reactphp/socket or https://github.com/amphp/socket for low level non-blocking IO operations

**Comment by lucasRolff** [comment]
I'd love something like batch-consumers that are available in your Symfony2 bundle: https://github.com/php-amqplib/RabbitMqBundle#batch-consumers

the use case I have is something like this:
- We have 100k incoming messages, all messages contain a string that has to be processed, a "string ID" and a callback URL
- Each string is processed, message after message
- Only wanna do a "callback" to the callback URL every X messages containing the "string ID" from every msg (since the callback system may not be able to handle the number of requests that we can process per second by the consumers).

I'm trying to find a nice way of doing it currently with php-amqplib directly but a bit lost.
